Context.if_enabled fails with AssertionError in interactive mode

Related #6129 

The program below fails in interactive mode 

```py
from Standard.Base import all
from Standard.Base.Runtime.Context import Output

main =
    res = Output.if_enabled 42
    res
```

The expression `Oubtput.if_enabled 42` should have an id assigned to be a subject of instrumentation. For a repro, see examples in the `RuntimeExecutionEnvironmentTest` suite.

<details>
<summary>

```
java.lang.AssertionError: Invariant contract violation for receiver Error Output(org.enso.interpreter.runtime.callable.atom.unboxing.Layout_Atom_0_1) and identifier to_display_text.
```

</summary>

```
	at org.graalvm.truffle/com.oracle.truffle.api.interop.InteropLibrary$Asserts.invokeMember(InteropLibrary.java:3462)
	at org.graalvm.truffle/com.oracle.truffle.api.interop.InteropLibraryGen$UncachedDispatch.invokeMember(InteropLibraryGen.java:7066)
	at org.enso.interpreter.service.ExecutionService.getExceptionMessage(ExecutionService.java:443)
	at org.enso.interpreter.instrument.job.ProgramExecutionSupport$.sendExpressionUpdate(ProgramExecutionSupport.scala:341)
	at org.enso.interpreter.instrument.job.ProgramExecutionSupport$.$anonfun$executeProgram$3(ProgramExecutionSupport.scala:74)
	at org.enso.interpreter.instrument.IdExecutionInstrument$IdExecutionEventNode.passExpressionValueToCallback(IdExecutionInstrument.java:301)
	at org.enso.interpreter.instrument.IdExecutionInstrument$IdExecutionEventNode.onExpressionReturn(IdExecutionInstrument.java:278)
	at org.enso.interpreter.instrument.IdExecutionInstrument$IdExecutionEventNode.onReturnValue(IdExecutionInstrument.java:237)
	at org.enso.interpreter.instrument.IdExecutionInstrument$IdExecutionEventNode.onReturnExceptional(IdExecutionInstrument.java:247)
	at org.graalvm.truffle/com.oracle.truffle.api.instrumentation.ProbeNode$EventProviderChainNode.innerOnReturnExceptional(ProbeNode.java:1387)
	at org.graalvm.truffle/com.oracle.truffle.api.instrumentation.ProbeNode$EventChainNode.onReturnExceptional(ProbeNode.java:1023)
	at org.graalvm.truffle/com.oracle.truffle.api.instrumentation.ProbeNode.onReturnExceptionalOrUnwind(ProbeNode.java:288)
	at org.enso.interpreter.node.ExpressionNodeWrapper.executeGeneric(ExpressionNodeWrapper.java:119)
	at org.enso.interpreter.node.scope.AssignmentNodeGen.executeGeneric_generic1(AssignmentNodeGen.java:57)
	at org.enso.interpreter.node.scope.AssignmentNodeGen.executeGeneric(AssignmentNodeGen.java:36)
	at org.enso.interpreter.node.scope.AssignmentNodeGen.executeVoid(AssignmentNodeGen.java:75)
	at org.enso.interpreter.node.callable.function.BlockNode.executeGeneric(BlockNode.java:54)
	at org.enso.interpreter.node.callable.function.BlockNode.executeGeneric(BlockNode.java:56)
	at org.enso.interpreter.node.ClosureRootNode.execute(ClosureRootNode.java:85)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.executeRootNode(OptimizedCallTarget.java:709)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.profiledPERoot(OptimizedCallTarget.java:632)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.callBoundary(OptimizedCallTarget.java:565)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.doInvoke(OptimizedCallTarget.java:549)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.callDirect(OptimizedCallTarget.java:495)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedDirectCallNode.call(OptimizedDirectCallNode.java:68)
	at org.enso.interpreter.node.callable.ExecuteCallNode.callDirect(ExecuteCallNode.java:56)
	at org.enso.interpreter.node.callable.ExecuteCallNodeGen.executeAndSpecialize(ExecuteCallNodeGen.java:90)
	at org.enso.interpreter.node.callable.ExecuteCallNodeGen.executeCall(ExecuteCallNodeGen.java:52)
	at org.enso.interpreter.node.callable.dispatch.SimpleCallOptimiserNode.executeDispatch(SimpleCallOptimiserNode.java:49)
	at org.enso.interpreter.node.callable.dispatch.CurryNode.doCall(CurryNode.java:157)
	at org.enso.interpreter.node.callable.dispatch.CurryNode.execute(CurryNode.java:108)
	at org.enso.interpreter.node.callable.dispatch.InvokeFunctionNode.invokeCached(InvokeFunctionNode.java:92)
	at org.enso.interpreter.node.callable.dispatch.InvokeFunctionNodeGen.executeAndSpecialize(InvokeFunctionNodeGen.java:102)
	at org.enso.interpreter.node.callable.dispatch.InvokeFunctionNodeGen.execute(InvokeFunctionNodeGen.java:58)
	at org.enso.interpreter.node.callable.InteropApplicationNode.callCached(InteropApplicationNode.java:79)
	at org.enso.interpreter.node.callable.InteropApplicationNodeGen.executeAndSpecialize(InteropApplicationNodeGen.java:95)
	at org.enso.interpreter.node.callable.InteropApplicationNodeGen.execute(InteropApplicationNodeGen.java:56)
	at org.enso.interpreter.node.callable.FunctionCallInstrumentationNode$FunctionCall$Execute.callCached(FunctionCallInstrumentationNode.java:103)
	at org.enso.interpreter.node.callable.FunctionCallGen$InteropLibraryExports$Cached.executeAndSpecialize(FunctionCallGen.java:91)
	at org.enso.interpreter.node.callable.FunctionCallGen$InteropLibraryExports$Cached.execute(FunctionCallGen.java:78)
	at org.graalvm.truffle/com.oracle.truffle.api.interop.InteropLibrary$Asserts.execute(InteropLibrary.java:3034)
	at org.graalvm.truffle/com.oracle.truffle.api.interop.InteropLibraryGen$CachedDispatch.execute(InteropLibraryGen.java:7765)
	at org.enso.interpreter.service.ExecutionService$ExecuteRootNode.execute(ExecutionService.java:477)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.executeRootNode(OptimizedCallTarget.java:709)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.profiledPERoot(OptimizedCallTarget.java:632)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.callBoundary(OptimizedCallTarget.java:565)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.doInvoke(OptimizedCallTarget.java:549)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.callIndirect(OptimizedCallTarget.java:477)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.call(OptimizedCallTarget.java:458)
	at org.enso.interpreter.service.ExecutionService.execute(ExecutionService.java:181)
	at org.enso.interpreter.service.ExecutionService.execute(ExecutionService.java:222)
	at org.enso.interpreter.instrument.job.ProgramExecutionSupport$.executeProgram(ProgramExecutionSupport.scala:106)
	at org.enso.interpreter.instrument.job.ProgramExecutionSupport$.$anonfun$runProgram$3(ProgramExecutionSupport.scala:215)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at cats.syntax.EitherObjectOps$.catchNonFatal$extension(either.scala:391)
	at org.enso.interpreter.instrument.job.ProgramExecutionSupport$.$anonfun$runProgram$2(ProgramExecutionSupport.scala:215)
	at scala.util.Either.flatMap(Either.scala:352)
	at org.enso.interpreter.instrument.job.ProgramExecutionSupport$.runProgram(ProgramExecutionSupport.scala:209)
	at org.enso.interpreter.instrument.job.ExecuteJob.run(ExecuteJob.scala:38)
	at org.enso.interpreter.instrument.job.ExecuteJob.run(ExecuteJob.scala:15)
	at org.enso.interpreter.instrument.execution.JobExecutionEngine.$anonfun$runInternal$1(JobExecutionEngine.scala:113)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread.access$001(PolyglotThread.java:53)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread$1.execute(PolyglotThread.java:100)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread$ThreadSpawnRootNode.executeImpl(PolyglotThread.java:134)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread$ThreadSpawnRootNode.execute(PolyglotThread.java:125)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.executeRootNode(OptimizedCallTarget.java:709)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.profiledPERoot(OptimizedCallTarget.java:632)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.callBoundary(OptimizedCallTarget.java:565)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.doInvoke(OptimizedCallTarget.java:549)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.callIndirect(OptimizedCallTarget.java:477)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.call(OptimizedCallTarget.java:458)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread.run(PolyglotThread.java:96)
Caused by: Attached Guest Language Frames (3)
```

</details>

--- a/engine/runtime-with-instruments/src/test/scala/org/enso/interpreter/test/instrument/RuntimeExecutionEnvironmentTest.scala
+++ b/engine/runtime-with-instruments/src/test/scala/org/enso/interpreter/test/instrument/RuntimeExecutionEnvironmentTest.scala
@@ -3,6 +3,7 @@ package org.enso.interpreter.test.instrument
 import org.enso.distribution.FileSystem
 import org.enso.distribution.locking.ThreadSafeFileLockManager
 import org.enso.interpreter.runtime.EnsoContext
+import org.enso.interpreter.runtime.`type`.ConstantsGen
 import org.enso.interpreter.test.Metadata
 import org.enso.pkg.{Package, PackageManager}
 import org.enso.polyglot._
@@ -29,6 +30,8 @@ class RuntimeExecutionEnvironmentTest
     with BeforeAndAfterEach
     with BeforeAndAfterAll
     with OsSpec {
+
+  import RuntimeExecutionEnvironmentTest.IF_ENABLED_METH_PTR
 
   override val timeLimit = 5.minutes
 
@@ -128,8 +131,6 @@ class RuntimeExecutionEnvironmentTest
 
   it should "panic when output context is not enabled" in {
 
-    pending
-
     val contextId  = UUID.randomUUID()
     val requestId  = UUID.randomUUID()
     val moduleName = "Enso_Test.Test.Main"
@@ -180,7 +181,10 @@ class RuntimeExecutionEnvironmentTest
       TestMessages.panic(
         contextId,
         idRes,
-        Api.ExpressionUpdate.Payload.Panic("Forbidden_Operation.Error", Seq())
+        IF_ENABLED_METH_PTR,
+        Api.ExpressionUpdate.Payload
+          .Panic("Forbidden operation: Output.", Seq(idRes)),
+        false
       ),
       context.executionComplete(contextId)
     )
@@ -199,8 +203,24 @@ class RuntimeExecutionEnvironmentTest
         )
       )
     )
-    context.receiveNIgnoreStdLib(2) should contain theSameElementsAs Seq(
+    context.receiveNIgnoreStdLib(3) should contain theSameElementsAs Seq(
       Api.Response(requestId, Api.SetExecutionEnvironmentResponse(contextId)),
+      Api.Response(
+        None,
+        Api.ExpressionUpdates(
+          contextId,
+          Set(
+            Api.ExpressionUpdate(
+              idRes,
+              Some(ConstantsGen.NOTHING),
+              Some(IF_ENABLED_METH_PTR),
+              Vector(Api.ProfilingInfo.ExecutionTime(0)),
+              false,
+              Api.ExpressionUpdate.Payload.Value()
+            )
+          )
+        )
+      ),
       context.executionComplete(contextId)
     )
     context.consumeOut shouldEqual List("Hello World!")
@@ -210,9 +230,6 @@ class RuntimeExecutionEnvironmentTest
   }
 
   it should "panic when input is not enabled" in {
-
-    pending
-
     val contextId  = UUID.randomUUID()
     val requestId  = UUID.randomUUID()
     val moduleName = "Enso_Test.Test.Main"
@@ -263,7 +280,10 @@ class RuntimeExecutionEnvironmentTest
       TestMessages.panic(
         contextId,
         idRes,
-        Api.ExpressionUpdate.Payload.Panic("Forbidden_Operation.Error", Seq())
+        IF_ENABLED_METH_PTR,
+        Api.ExpressionUpdate.Payload
+          .Panic("Forbidden operation: Input.", Seq(idRes)),
+        false
       ),
       context.executionComplete(contextId)
     )
@@ -282,8 +302,24 @@ class RuntimeExecutionEnvironmentTest
         )
       )
     )
-    context.receiveNIgnoreStdLib(2) should contain theSameElementsAs Seq(
+    context.receiveNIgnoreStdLib(3) should contain theSameElementsAs Seq(
       Api.Response(requestId, Api.SetExecutionEnvironmentResponse(contextId)),
+      Api.Response(
+        None,
+        Api.ExpressionUpdates(
+          contextId,
+          Set(
+            Api.ExpressionUpdate(
+              idRes,
+              Some(ConstantsGen.INTEGER),
+              Some(IF_ENABLED_METH_PTR),
+              Vector(Api.ProfilingInfo.ExecutionTime(0)),
+              false,
+              Api.ExpressionUpdate.Payload.Value()
+            )
+          )
+        )
+      ),
       context.executionComplete(contextId)
     )
     context.languageContext.getExecutionEnvironment.getName shouldEqual Api.ExecutionEnvironment
@@ -291,4 +327,12 @@ class RuntimeExecutionEnvironmentTest
       .name
   }
 
+}
+
+object RuntimeExecutionEnvironmentTest {
+  val IF_ENABLED_METH_PTR = Api.MethodPointer(
+    "Standard.Base.Runtime",
+    "Standard.Base.Runtime.Context",
+    "if_enabled"
+  )
 }

--- a/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/error/displaytext/ForbiddenOperationToDisplayTextNode.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/error/displaytext/ForbiddenOperationToDisplayTextNode.java
@@ -1,0 +1,31 @@
+package org.enso.interpreter.node.expression.builtin.error.displaytext;
+
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.nodes.Node;
+import org.enso.interpreter.dsl.BuiltinMethod;
+import org.enso.interpreter.runtime.callable.atom.Atom;
+import org.enso.interpreter.runtime.callable.atom.AtomConstructor;
+import org.enso.interpreter.runtime.callable.atom.StructsLibrary;
+import org.enso.interpreter.runtime.data.text.Text;
+
+@BuiltinMethod(type = "Forbidden_Operation", name = "to_display_text")
+public abstract class ForbiddenOperationToDisplayTextNode extends Node {
+  static ForbiddenOperationToDisplayTextNode build() {
+    return ForbiddenOperationToDisplayTextNodeGen.create();
+  }
+
+  abstract Text execute(Object self);
+
+  @Specialization
+  Text doAtom(Atom self, @CachedLibrary(limit = "3") StructsLibrary structs) {
+    return Text.create("Forbidden operation: ")
+        .add(String.valueOf(structs.getField(self, 0)))
+        .add(".");
+  }
+
+  @Specialization
+  Text doConstructor(AtomConstructor self) {
+    return Text.create("Forbidden operation.");
+  }
+}

--- a/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/error/displaytext/ForbiddenOperationToDisplayTextNode.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/error/displaytext/ForbiddenOperationToDisplayTextNode.java
@@ -1,5 +1,6 @@
 package org.enso.interpreter.node.expression.builtin.error.displaytext;
 
+import com.oracle.truffle.api.CompilerDirectives;
 import com.oracle.truffle.api.dsl.Specialization;
 import com.oracle.truffle.api.library.CachedLibrary;
 import com.oracle.truffle.api.nodes.Node;
@@ -18,6 +19,7 @@ public abstract class ForbiddenOperationToDisplayTextNode extends Node {
   abstract Text execute(Object self);
 
   @Specialization
+  @CompilerDirectives.TruffleBoundary
   Text doAtom(Atom self, @CachedLibrary(limit = "3") StructsLibrary structs) {
     return Text.create("Forbidden operation: ")
         .add(String.valueOf(structs.getField(self, 0)))

